gui/mozregui/report.py

```diff
@@ -73,22 +96,48 @@ def clear(self):
 
     @Slot(object)
     def attach_bisector(self, bisector):
-        slots = ('step_started', 'step_build_found', 'step_testing',
-                 'step_finished', 'started', 'finished')
+        bisector_slots = ('step_started',
+                          'step_build_found',
+                          'step_testing',
+                          'step_finished',
+                          'started',
+                          'finished')
+        downloader_slots = ('download_progress',
+                            'download_started',
+                            'download_finished')
+
         if self.bisector:
-            # disconnect previous bisector
-            for name in slots:
-                signal = getattr(self.bisector, name)
-                slot = getattr(self, name)
-                signal.disconnect(slot)
+            bulk_disconnect_slots(bisector_slots, self.bisector, self)
+            bulk_disconnect_slots(downloader_slots,
+                                  self.bisector.download_manager,
+                                  self)
+
         if bisector:
-            # connect the new bisector
-            for name in slots:
-                signal = getattr(bisector, name)
-                slot = getattr(self, name)
-                signal.connect(slot)
+            bulk_connect_slots(bisector_slots, bisector, self)
+            bulk_connect_slots(downloader_slots,
+                               bisector.download_manager,
+                               self)
+
         self.bisector = bisector
 
+    @Slot(object, int, int)
+    def download_progress(self, dl, current, total):
+        item = self.items[-1]
+        item.progress = (current * 100) / total
+        index = self.createIndex(len(self.items) - 1, 0)
+        self.dataChanged.emit(index, index)
+
+    @Slot(object)
+    def download_started(self):
+        self.is_downloading = True
+
+    @Slot(object, str)
+    def download_finished(self, fname):
+        self.is_downloading = False
+
+    def last_item(self):
```

Comment:
not sure if this is useful. If you want to avoid accessing the items variables from outside, maybe just implement an item method. you could make it so that last_item() == item(-1). This would make the function more useful.